### ISSUE TITLE
Support PIE Models

### ISSUE THREAD

**Issue by thomaspinder** [pymc-labs/pymc-marketing#2507]
Proposal herein for how we can offer support for the Predicted Incrementality by Experimentation (PIE) of Gordon, Moakler & Zettelmeyer (2026). This issue proposes a minimum-viable `pymc_marketing.pie` submodule that ports PIE to pymc-marketing with a Bayesian predictive core (PyMC-BART). It is intentionally scoped below the paper's full offering; the decision-theoretic framework and MMM-integration variants are left as follow-ups and called out under "Alternatives considered".

### In scope (MVP)

- A `pymc_marketing/pie/` submodule with a `PIEModel` class that inherits from `ModelBuilder`.
- Training API: `PIEModel.build_model(X, y)` / `fit(X, y)` where `X` is a dataframe of campaign-level features from a corpus of past RCTs and `y` is the measured causal effect (default: incremental conversions per dollar).
- Prediction API: `sample_posterior_predictive(X_new)` returning an `xarray` with posterior draws of predicted incrementality for non-experimental campaigns.
- Support for both **pre-determined** features (objective, vertical, budget, audience type) and **post-determined** features (exposure rate, CTR, last-click conversions, average treated outcome), per §3 of the paper.
- A synthetic RCT generator in `pymc_marketing/pie/synthetic.py` that simulates a corpus of campaigns with known heterogeneous treatment effects, used by tests and the example notebook.
- A documentation notebook under `docs/source/notebooks/pie/` demonstrating: simulate → fit → predict → compare against a last-click-style baseline.
- Serialization / load, MLflow autologging, and `default_model_config`-based prior overrides inherited from `ModelBuilder`.

### Out of scope (explicit — follow-ups)

- Decision-theoretic framework (§6 of paper): go/no-go thresholds, Type I / Type II error accounting, disagreement-rate metrics against an RCT ground truth.
    - Perhaps we can offer a small demo of this in a documentation notebook. Alternatively, it can become part of the core library in a subsequent PR. Doing it here will bloat the PR though.
- Extrapolation / subgroup diagnostics (§5.3): accuracy degradation when the target campaign's segment is absent from training.
- PIE → MMM integration: using PIE predictions as informative priors on MMM saturation / ROI parameters.
- Pluggable non-BART backends (hierarchical Bayesian regression, GPs).

## Module layout

```
pymc_marketing/pie/
    __init__.py           # re-exports PIEModel, generate_synthetic_rct_corpus
    pie.py                # PIEModel(ModelBuilder)
    features.py           # feature schema, pre/post classification, validation
    synthetic.py          # generate_synthetic_rct_corpus(...)
tests/pie/
    test_pie_model.py     # fit / predict / save-load / prior override
    test_synthetic.py     # recovery on synthetic corpus with known effects
docs/source/notebooks/pie/
    pie_quickstart.ipynb
```

`pymc-bart` is declared as an optional extra in `pyproject.toml` (`pymc-marketing[pie]`). `pie.py` uses a lazy import and raises a clear `ImportError` with install instructions if it is missing.

## API sketch

```python
import pandas as pd
from pymc_extras.prior import Prior
from pymc_marketing.pie import PIEModel, generate_synthetic_rct_corpus

rct_corpus = generate_synthetic_rct_corpus(n_campaigns=500, seed=42)
# rct_corpus columns: campaign_id, objective, vertical, budget, audience_type,
#                     exposure_rate, ctr, last_click_conversions_per_dollar,
#                     avg_treated_outcome, measured_incrementality_per_dollar

X = rct_corpus.drop(columns=["campaign_id", "measured_incrementality_per_dollar"])
y = rct_corpus["measured_incrementality_per_dollar"]

model = PIEModel(
    pre_determined_features=["objective", "vertical", "budget", "audience_type"],
    post_determined_features=["exposure_rate", "ctr",
                              "last_click_conversions_per_dollar",
                              "avg_treated_outcome"],
    target="measured_incrementality_per_dollar",
    model_config={
        "bart": {"m": 200, "alpha": 0.95, "beta": 2.0},
        "sigma": Prior("HalfNormal", sigma=1.0),
    },
)

model.build_model(X, y)
model.fit(X, y, nuts_sampler="numpyro", random_seed=42)

# Predict for non-experimental campaigns (same schema, no y required)
preds = model.sample_posterior_predictive(X_new)  # xarray with chain/draw dims
```

### Key design notes

- `PIEModel.__init__` validates inputs with pydantic and stores `model_config`; the PyMC graph is built in `build_model`, matching the house lifecycle.
- `default_model_config` exposes the BART hyperparameters (`m`, `alpha`, `beta`) and noise prior as overridable `Prior` objects.
- Categorical features are handled by BART's native integer-split logic after label-encoding inside `features.py`.
- `pre_determined_features` vs. `post_determined_features` are tracked separately on the class so future work can gate on availability (e.g., predict-at-launch with only pre-determined features).

### Training row (one row per past RCT campaign)

| Column | Type | Role | Notes |
|---|---|---|---|
| `campaign_id` | str/int | id | not a feature |
| `objective` | categorical | pre-determined | e.g. `conversions`, `traffic`, `awareness` |
| `vertical` | categorical | pre-determined | e.g. `retail`, `travel`, `finance` |
| `audience_type` | categorical | pre-determined | e.g. `prospecting`, `retargeting` |
| `budget` | float ≥ 0 | pre-determined | total campaign spend |
| `exposure_rate` | float ∈ [0, 1] | post-determined | fraction of the targeted group actually exposed |
| `ctr` | float ∈ [0, 1] | post-determined | click-through rate in the treated arm |
| `last_click_conversions_per_dollar` | float ≥ 0 | post-determined | attributed-metric benchmark (also used as a comparison baseline) |
| `avg_treated_outcome` | float | post-determined | mean outcome in the treated arm |
| `measured_incrementality_per_dollar` | float | **target `y`** | `(treated_mean − control_mean) / spend`, from the RCT |

The schema is intentionally close to the paper's §4 description and extensible: users can add additional pre- or post-determined feature columns and pass them via the `pre_determined_features` / `post_determined_features` lists.

### Prediction row (one row per non-experimental campaign)

Same columns **except** `measured_incrementality_per_dollar` (the quantity being predicted). `control_mean` is not required for prediction — that is the whole point of PIE.

### Validation (in `features.py`)

- `pre_determined_features ∪ post_determined_features ⊆ X.columns`.
- At prediction time, `X_new` must contain all columns the model was trained on, with the same categorical levels. Unseen categorical levels raise a clear error (follow-up: smoothing / fallback to marginal mean).
- Rates (`exposure_rate`, `ctr`) bounded in `[0, 1]`; `budget` and per-dollar targets must be non-negative.

### Likelihood (v1)

```
μ(x) = BART(x; m, α, β)            # sum of regression trees on all features
y ~ Normal(μ(x), σ)                # σ ~ HalfNormal(1.0)
```

The BART component is `pymc_bart.BART("bart", X, Y=y, m=200, alpha=0.95, beta=2.0)`, added directly inside the `pm.Model` context, following `pymc-bart`'s standard API. Categorical features are handled by BART's native split logic after integer-encoding.

### Known limitations flagged up front

- `pm.sample` with BART is slower than the paper's random forest; on a 500-campaign corpus the fit should complete in minutes on CPU.
- BART's posterior uncertainty captures model uncertainty but not the sampling variance in the RCT estimates themselves (footnote 2 of the paper). A follow-up could add a measurement-error layer `y_observed ~ Normal(y_true, se_rct)` if users supply per-campaign standard errors.

### Tests (`tests/pie/`)

- `test_synthetic.py` — `generate_synthetic_rct_corpus` returns a dataframe with the documented schema and tunable heterogeneity.
- `test_pie_model.py`:
  - `build_model` produces a valid PyMC model with expected variables / coords.
  - `fit` → `sample_posterior_predictive` recovers known treatment effects within tolerance on synthetic data (parametrized over corpus sizes 100 / 500).
  - `save` / `load` round-trip preserves posterior and config.
  - `model_config` override changes BART hyperparameters as expected.
  - Unseen categorical level at predict time raises a clear error.

### Example notebook (`docs/source/notebooks/pie/pie_quickstart.ipynb`)

1. Simulate an RCT corpus with `generate_synthetic_rct_corpus`.
2. Fit `PIEModel`.
3. Predict on a held-out set of "non-experimental" campaigns.
4. Compare predicted vs. true incrementality; compare PIE's R² against a last-click baseline (`last_click_conversions_per_dollar` as the prediction), reproducing the paper's headline qualitative result on synthetic data.

## Alternatives considered

- **Full Bayesian PIE in one PR** — include the paper's decision framework (go/no-go thresholds, Type I/II error accounting, disagreement-rate vs. RCT) and subgroup/extrapolation diagnostics (§5.3) alongside the predictive core. Rejected for the MVP on scope grounds; happy to land it immediately after if maintainers prefer a larger first landing.
- **PIE as an MMM extension / calibration layer** — instead of a standalone predictor, use a corpus of external RCTs to inform priors on an MMM's saturation/ROI parameters, next to `add_lift_test_measurements`. Different problem shape (per-channel curve calibration vs. cross-campaign effect prediction); worth pursuing separately if there is appetite, but conflating it with PIE v1 muddies both.
- **Pluggable backend** — `PIEModel` as an abstract base with `BARTPIEModel`, `HierarchicalPIEModel`, etc. Deferred; land one concrete model first, generalize if a second backend is actually requested.
- **Hierarchical Bayesian regression instead of BART** — more interpretable and slots more naturally into pymc-marketing's `Prior` idioms, but gives up the nonparametric flexibility the paper relied on. Viable as a second backend under the pluggable-backend extension.
- **Measurement-error layer** — `y_obs ~ Normal(y_true, se_rct)` to propagate per-campaign RCT uncertainty. Deferred; adds API surface (users must supply SEs) for a secondary effect.

## Open questions for maintainers

1. **Home of the module**: `pymc_marketing/pie/` (proposed) or nested under `pymc_marketing/mmm/pie.py`? The paper frames PIE as an MMM alternative, not an extension, which argues for a sibling module.
2. **Target default**: the paper centers on incremental conversions per dollar. Should the API default to this, or be target-agnostic (users always pass the target column name)?
3. **`pymc-bart` as optional extra** — confirming this matches the project's dependency policy. The alternative is a hard dependency for simpler UX.
4. **Categorical handling**: BART's native integer-split vs. one-hot encoding upstream? Impacts interpretability of variable-importance outputs.
5. **Naming**: `PIEModel` is terse but accurate. Alternatives: `PredictedIncrementalityModel`, `CrossCampaignIncrementality`.
6. **Appetite for follow-ups**: is there interest in the decision framework (§6) and extrapolation diagnostics (§5.3) as immediate follow-up issues, or should we see how v1 lands first?

**Comment by juanitorduz** [comment]
This is great! Thanks @thomaspinder ! Regarding your points:

1. ` pymc_marketing/mmm/pie.py` should be a good start.
2. Ideally, target agnostic.
3. We can have it as an optional dependency, see https://github.com/pymc-labs/pymc-marketing/blob/main/pyproject.toml#L50
4. I do not think there is a silver bullet, depends on the variable, let me think about it (we can start with either or)
5. No, strong opinions. I like `PIEModel`.
6. Definitively! Let's just work on iterations :)
